pkg/scaffold/resource/options.go

```diff
@@ -0,0 +1,231 @@
+
+// Options contains the information required to build a new Resource
+type Options struct {
+	// Namespaced is true if the resource is namespaced.
+	Namespaced bool
+
+	// Group is the API Group. Does not contain the domain.
+	Group string
+
+	// Version is the API version.
+	Version string
+
+	// Kind is the API Kind.
+	Kind string
+
+	// Plural is the API Kind plural form.
+	// Optional
+	Plural string
+
+	// Resource is the API Resource.
+	// Optional
+	Resource string
+
+	// CreateExampleReconcileBody will create a Deployment in the Reconcile example.
+	// v1 only
+	CreateExampleReconcileBody bool
+}
+
+func (opts *Options) Validate() error {
+	// Check that the required flags did not get a flag as their value
+	// NOTE(adirio): We must do this for all the required flags first or we may
+	// output the wrong error as flags may seem to be missing because Cobra
+	// assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupRequired)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionRequired)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindRequired)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Group) == 0 {
+		return fmt.Errorf(groupRequired)
+	}
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the Group has a valid DNS1123 subdomain value
+	if err := IsDNS1123Subdomain(opts.Group); err != nil {
+		return fmt.Errorf("group name is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf(
+			"version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	// Check if the Kind is PascalCase
+	if opts.Kind != flect.Pascalize(opts.Kind) {
+		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(opts.Kind), opts.Kind)
+	}
+
+	// TODO: validate plural and resource strings if provided
+
+	return nil
+}
+
+// safeImport returns a cleaned version of the provided string that can be used for imports
+func (opts *Options) safeImport(unsafe string) string {
+	safe := unsafe
+
+	// Remove dashes and dots
+	safe = strings.Replace(safe, "-", "", -1)
+	safe = strings.Replace(safe, ".", "", -1)
+
+	return safe
+}
+
+func (opts *Options) NewV1Resource(repo, domain string) *Resource {
+	res := opts.newResource()
+
+	// Use the k8s.io/api package for core resources
+	coreGroups := map[string]string{
+		"apps":                  "",
+		"admissionregistration": "k8s.io",
+		"apiextensions":         "k8s.io",
+		"authentication":        "k8s.io",
+		"autoscaling":           "",
+		"batch":                 "",
+		"certificates":          "k8s.io",
+		"core":                  "",
+		"extensions":            "",
+		"metrics":               "k8s.io",
+		//"policy":                "",
```

Comment:
This commented line is conflictive. If uncommented, tests will fail. I will post my understanding about this, some help would be apreciated, feel free to contact me through kubebuilder slack channel.

The behavior (both before and after this PR) is to check if the provided resource had the expected type file. In case it doesn't, check if it is one of the builtin apis, and change the domain and package import path accordingly. This allows to have custom api groups with the same name than builting api groups. In case neither the file nor the builtin api group is found, it defaults to consider it local.

When the golden test is building the testdata folder, it seems to be unable to find this file, so it tries to look for a builtin group. The resources in the `crew` group do not find a builtin api group so they default back to local files. The resources in `core` group do not find the file and default to the builtin "core" api group (as expected). But the resources in the `policy` group do not find the file and thus look for a bultin api group, which they would find without this line commented, and thus consider the resources in this group part of the builtin "policy" api group when it shouldn't.

The thing is that checking this file was already being done, the behavior hasn't changed, just moved to this file.

When creating the testdata to check the expected files, the `os.Stat` call in line 149 is not finding the file. This check was already being done previously. In case you cant find the types definition file, you look for the builtin groups, but you prioritize the custom types in case you do find the mentioned file.

Reply:
It's unclear to me why this was commented out in the first place, though.  Policy is definitely a core API group (FWIW, this is why we have the whole subdomain thing in core kubernetes now).

Reply:
First, let me be supremely unhelpful: I don't like this logic in the first place.  We should have a syntax for explicitly specifying external API groups, and use it to disambiguate.

Ok, now for what to do about this: if the issue is actually that we're in testdata, we might need to `Chdir` into the testdata directory in the file first, or something.  testdata gets ignored by go when you're outside of it.  Let me try to reproduce.

Reply:
I strongly agree with that. Like a --builtin flag or something similar.

Reply:
Aaaah.  Our scaffolding is just a mess (but you already knew that).  Before your refactor, for v1, we'd do the check and build the universe with `policy` treated as core.

Then, we'd *completely* ignore that, generate the API anyway, and then *re-run that check in a different method* when generating the controller.  The second time we run the check, API exists, so the file exists, so we treat policy as local.

With your change, AFAICT, we just run the check once, so we treat policy as core.

Reply:
:woman_facepalming:

Reply:
(your version is sane, FWIW)

As a follow-up: all of this "check if the file exists" logic should probably go away eventually in favor of:

- if the API is NON-fully qualified, treat as local (e.g. `policy`, `metrics`, etc)
- if the API is fully qualified, check the PROJECT file (e.g. `policy.` (note the trailing period), `metrics.k8s.io`):
  - if it exists in the PROJECT file, treat as local
  - if it doesn't exist in the PROJECT file, treat as external

Reply:
For now, I think we just need to special-case v1 to *always* consider the file as existing when generating *both* API and *controller*.

Reply:
> As a follow-up: all of this "check if the file exists" logic should probably go away eventually in favor of:
> 
> * if the API is NON-fully qualified, treat as local (e.g. `policy`, `metrics`, etc)
> * if the API is fully qualified, check the PROJECT file (e.g. `policy.` (note the trailing period), `metrics.k8s.io`):
>   
>   * if it exists in the PROJECT file, treat as local
>   * if it doesn't exist in the PROJECT file, treat as external

What would you say that are the requirements to consider an API fully qualified? GVK is enforced, and domain is obtained from the PROJECT file. Should there be a optional domain flag for `kubebuilder create api ...` that defaults to the PROJECT domain? In this case, by using "k8s.io" as the API domain the scaffolder will know that is one of the core types.

Reply:
> Aaaah. Our scaffolding is just a mess (but you already knew that). Before your refactor, for v1, we'd do the check and build the universe with `policy` treated as core.
> 
> Then, we'd _completely_ ignore that, generate the API anyway, and then _re-run that check in a different method_ when generating the controller. The second time we run the check, API exists, so the file exists, so we treat policy as local.
> 
> With your change, AFAICT, we just run the check once, so we treat policy as core.

I recreated the resources after scaffolding the resource type files as a temporary solution. I did it for both v1 and v2 (policy was not giving any issues for v2 because it is commented in generate_golden.sh due to v2 being single-group only but v2 multi-group is coming soon).